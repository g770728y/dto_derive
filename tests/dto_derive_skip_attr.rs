use dto_derive::Dto;

#[derive(Debug, PartialEq)]
struct Entity {
    pub a: String,
    pub b: i32,
    pub c: bool,
}

#[derive(Debug, PartialEq, Dto)]
#[dto(entity = "Entity")]
#[dto(skip = "d")]
struct DtoSkipRequest {
    a: String,
    b: i32,
    c: bool,
    d: String,
}

#[derive(Debug, PartialEq, Dto)]
#[dto(entity = "Entity")]
#[dto(skip = "d,e,f")]
struct DtoMultiSkipOnceRequest {
    a: String,
    b: i32,
    c: bool,
    d: i32,
    e: i32,
    f: i32,
}

#[derive(Debug, PartialEq, Dto)]
#[dto(entity = "Entity")]
#[dto(skip = "d")]
#[dto(skip = "e")]
#[dto(skip = "f")]
struct DtoMultiSkipRequest {
    a: String,
    b: i32,
    c: bool,
    d: i32,
    e: i32,
    f: i32,
}

#[test]
fn dto_request_one_skip_attr() {
    let request = DtoSkipRequest {
        a: "test_simple".to_string(),
        b: 123i32,
        c: true,
        d: "skipped".to_string(),
    };

    let expected = Entity {
        a: "test_simple".to_string(),
        b: 123i32,
        c: true,
    };

    let actual: Entity = request.into();
    assert_eq!(expected, actual);
}

#[test]
fn dto_request_multi_skip_once_attr() {
    let request = DtoMultiSkipOnceRequest {
        a: "test_simple".to_string(),
        b: 123i32,
        c: true,
        d: 10i32,
        e: 10i32,
        f: 10i32,
    };

    let expected = Entity {
        a: "test_simple".to_string(),
        b: 123i32,
        c: true,
    };

    let actual: Entity = request.into();
    assert_eq!(expected, actual);
}

#[test]
fn dto_request_multi_skip_attr() {
    let request = DtoMultiSkipRequest {
        a: "test_simple".to_string(),
        b: 123i32,
        c: true,
        d: 10i32,
        e: 10i32,
        f: 10i32,
    };

    let expected = Entity {
        a: "test_simple".to_string(),
        b: 123i32,
        c: true,
    };

    let actual: Entity = request.into();
    assert_eq!(expected, actual);
}
