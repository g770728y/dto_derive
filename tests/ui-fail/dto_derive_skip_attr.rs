extern crate dto_derive;
use dto_derive::Dto;

#[derive(Debug, PartialEq, Dto)]
#[dto(skip = 123)] //~ ERROR expected string literal containing field names separated by comma
struct Dto1 { }

#[derive(Debug, PartialEq, Dto)]
#[dto(skip = "")] //~ ERROR expected at least one field name
struct Dto2 { }

#[derive(Debug, PartialEq, Dto)]
#[dto(skip = "a, b, c")] //~ ERROR cannot skip non-existent field 'a'
struct Dto3 { }

#[derive(Debug, PartialEq, Dto)]
#[dto(skip = "a, b, c")] //~ ERROR cannot skip non-existent field 'b'
struct Dto4 {
    a: String,
}

#[derive(Debug, PartialEq, Dto)]
#[dto(skip = "a, b, c")]
#[dto(skip = "b")] //~ ERROR cannot skip already skipped field 'b'
struct Dto5 {
    a: String,
    b: String,
    c: String,
}

#[derive(Debug, PartialEq, Dto)]
#[dto(skip = "a, b, b, c")]  //~ ERROR cannot skip already skipped field 'b'
struct Dto6 {
    a: String,
    b: String,
    c: String,
}

fn main() { }
