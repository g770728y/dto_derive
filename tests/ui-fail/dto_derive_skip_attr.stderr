error: expected string literal containing field names separated by comma
 --> $DIR/dto_derive_skip_attr.rs:5:14
  |
5 | #[dto(skip = 123)] //~ ERROR expected string literal containing field names separated by comma
  |              ^^^

error: expected at least one field name
 --> $DIR/dto_derive_skip_attr.rs:9:14
  |
9 | #[dto(skip = "")] //~ ERROR expected at least one field name
  |              ^^

error: cannot skip non-existent field 'a'
  --> $DIR/dto_derive_skip_attr.rs:13:6
   |
13 | #[dto(skip = "a, b, c")] //~ ERROR cannot skip non-existent field 'a'
   |      ^^^^^^^^^^^^^^^^^^

error: cannot skip non-existent field 'b'
  --> $DIR/dto_derive_skip_attr.rs:17:6
   |
17 | #[dto(skip = "a, b, c")] //~ ERROR cannot skip non-existent field 'b'
   |      ^^^^^^^^^^^^^^^^^^

error: cannot skip already skipped field 'b'
  --> $DIR/dto_derive_skip_attr.rs:24:6
   |
24 | #[dto(skip = "b")] //~ ERROR cannot skip already skipped field 'b'
   |      ^^^^^^^^^^^^

error: cannot skip already skipped field 'b'
  --> $DIR/dto_derive_skip_attr.rs:32:6
   |
32 | #[dto(skip = "a, b, b, c")]  //~ ERROR cannot skip already skipped field 'b'
   |      ^^^^^^^^^^^^^^^^^^^^^

error: aborting due to 6 previous errors

